File: generated_code/cpp/solution_432_1.cpp
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <list>

using namespace std;

class AllOne {
private:
    // Node structure for the doubly linked list.
    // Each node represents a specific frequency (count) and stores all keys with that frequency.
    struct Node {
        int count;
        unordered_set<string> keys;
        Node(int c) : count(c) {}
    };

    // Doubly linked list to maintain sorted order of frequencies.
    list<Node> bucketList;
    
    // Maps each key to its corresponding iterator in the bucketList.
    unordered_map<string, list<Node>::iterator> keyToBucket;

public:
    AllOne() {
        // Initialize the object. The bucketList is initially empty.
    }
    
    void inc(string key) {
        if (keyToBucket.find(key) == keyToBucket.end()) {
            // Case 1: Key does not exist. It needs to be inserted with count 1.
            if (bucketList.empty() || bucketList.front().count != 1) {
                bucketList.push_front(Node(1));
            }
            bucketList.front().keys.insert(key);
            keyToBucket[key] = bucketList.begin();
        } else {
            // Case 2: Key exists. Move it to the bucket with count + 1.
            auto curBucket = keyToBucket[key];
            auto nextBucket = next(curBucket);
            int nextCount = curBucket->count + 1;
            
            // Create a new bucket if the bucket with nextCount doesn't exist.
            if (nextBucket == bucketList.end() || nextBucket->count != nextCount) {
                nextBucket = bucketList.insert(nextBucket, Node(nextCount));
            }
            
            nextBucket->keys.insert(key);
            keyToBucket[key] = nextBucket;
            
            // Remove the key from the old bucket.
            curBucket->keys.erase(key);
            if (curBucket->keys.empty()) {
                bucketList.erase(curBucket);
            }
        }
    }
    
    void dec(string key) {
        // It is guaranteed that key exists in the data structure before decrement.
        auto curBucket = keyToBucket[key];
        int nextCount = curBucket->count - 1;
        
        if (nextCount == 0) {
            // If count becomes 0, completely remove the key.
            keyToBucket.erase(key);
        } else {
            // Move the key to the bucket with count - 1.
            auto prevBucket = curBucket;
            if (prevBucket == bucketList.begin()) {
                bucketList.push_front(Node(nextCount));
                prevBucket = bucketList.begin();
            } else {
                prevBucket = prev(curBucket);
                if (prevBucket->count != nextCount) {
                    prevBucket = bucketList.insert(curBucket, Node(nextCount));
                }
            }
            
            prevBucket->keys.insert(key);
            keyToBucket[key] = prevBucket;
        }
        
        // Remove the key from the old bucket.
        curBucket->keys.erase(key);
        if (curBucket->keys.empty()) {
            bucketList.erase(curBucket);
        }
    }
    
    string getMaxKey() {
        // The last node in bucketList contains keys with the maximum frequency.
        if (bucketList.empty()) {
            return "";
        }
        return *(bucketList.back().keys.begin());
    }
    
    string getMinKey() {
        // The first node in bucketList contains keys with the minimum frequency.
        if (bucketList.empty()) {
            return "";
        }
        return *(bucketList.front().keys.begin());
    }
};

/**
 * Your AllOne object will be instantiated and called as such:
 * AllOne* obj = new AllOne();
 * obj->inc(key);
 * obj->dec(key);
 * string param_3 = obj->getMaxKey();
 * string param_4 = obj->getMinKey();
 */

File: generated_code/python/solution_432_1.py
from typing import Dict, Set

class Node:
    """
    A node in the doubly linked list.
    Each node represents a specific frequency/count and contains all keys with that count.
    """
    def __init__(self, count: int):
        self.count: int = count
        self.keys: Set[str] = set()
        self.prev: 'Node' = None
        self.next: 'Node' = None

class AllOne:
    """
    Design a data structure supporting inc, dec, getMaxKey, getMinKey in O(1) average time.
    Algorithm: Doubly Linked List of frequency nodes + Hash Map mapping keys to their nodes.
    """
    def __init__(self):
        # Maps each key to its corresponding Node in the doubly linked list
        self.key_to_node: Dict[str, Node] = {}
        
        # Sentinel head and tail for the doubly linked list to avoid edge cases
        self.head: Node = Node(0)
        self.tail: Node = Node(0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def _insert_node_after(self, new_node: Node, base_node: Node) -> None:
        """Helper to insert new_node right after base_node."""
        new_node.prev = base_node
        new_node.next = base_node.next
        base_node.next.prev = new_node
        base_node.next = new_node

    def _remove_node_if_empty(self, node: Node) -> None:
        """Helper to remove a node from the linked list if it contains no keys."""
        if not node.keys:
            node.prev.next = node.next
            node.next.prev = node.prev

    def inc(self, key: str) -> None:
        if key not in self.key_to_node:
            # Key is new, initial count should be 1
            first_node = self.head.next
            if first_node == self.tail or first_node.count > 1:
                # Node with count 1 doesn't exist, create it
                new_node = Node(1)
                self._insert_node_after(new_node, self.head)
                first_node = new_node
            
            first_node.keys.add(key)
            self.key_to_node[key] = first_node
        else:
            # Key exists, increment its count
            curr_node = self.key_to_node[key]
            next_node = curr_node.next
            
            if next_node == self.tail or next_node.count > curr_node.count + 1:
                # Node with count + 1 doesn't exist, create it
                new_node = Node(curr_node.count + 1)
                self._insert_node_after(new_node, curr_node)
                next_node = new_node
                
            next_node.keys.add(key)
            self.key_to_node[key] = next_node
            
            curr_node.keys.remove(key)
            self._remove_node_if_empty(curr_node)

    def dec(self, key: str) -> None:
        # It is guaranteed that key exists before decrementing
        curr_node = self.key_to_node[key]
        prev_node = curr_node.prev
        
        if curr_node.count == 1:
            # If count becomes 0, remove the key entirely
            del self.key_to_node[key]
        else:
            if prev_node == self.head or prev_node.count < curr_node.count - 1:
                # Node with count - 1 doesn't exist, create it
                new_node = Node(curr_node.count - 1)
                self._insert_node_after(new_node, prev_node)
                prev_node = new_node
                
            prev_node.keys.add(key)
            self.key_to_node[key] = prev_node
            
        curr_node.keys.remove(key)
        self._remove_node_if_empty(curr_node)

    def getMaxKey(self) -> str:
        # The maximal count node is always just before the tail sentinel
        if self.tail.prev == self.head:
            return ""
        # Return any key from the set
        return next(iter(self.tail.prev.keys))

    def getMinKey(self) -> str:
        # The minimal count node is always just after the head sentinel
        if self.head.next == self.tail:
            return ""
        # Return any key from the set
        return next(iter(self.head.next.keys))